src/platforms/hosted/bmp_libusb.c

```diff
@@ -47,291 +106,373 @@ void libusb_exit_function(bmp_info_s *info)
 	}
 }
 
-static bmp_type_t find_cmsis_dap_interface(libusb_device *dev, bmp_info_s *info)
+char *get_device_descriptor_string(libusb_device_handle *handle, uint16_t string_index)
 {
-	bmp_type_t type = BMP_TYPE_NONE;
+	char read_string[128] = {0};
+	const int length =
+		libusb_get_string_descriptor_ascii(handle, string_index, (unsigned char *)read_string, sizeof(read_string));
+	if (length <= 0) {
+		DEBUG_WARN("libusb_get_string_descriptor() failed with: %s (%d)\n", libusb_error_name(length), length);
+		return NULL;
+	} else if (!length)
+		read_string[0] = '\0';
+	return strdup(read_string);
+}
 
-	libusb_config_descriptor_s *conf;
-	char interface_string[128];
+//
+// BMP Probes have their version information in the product string.
+//
+// Extract the product and version, skip the manufacturer string
+//
+void get_bmp_product_version_string(struct libusb_device_descriptor *device_descriptor, libusb_device *device,
+	libusb_device_handle *handle, char **product, char **manufacturer, char **serial, char **version)
+{
+	(void)device;
+	(void)serial;
+	(void)manufacturer;
+	char *start_of_version;
+	*product = get_device_descriptor_string(handle, device_descriptor->iProduct);
+	start_of_version = strrchr(*product, ' ');
+	if (start_of_version == NULL) {
+		version = NULL;
+	} else {
+		while (start_of_version[0] == ' ' && start_of_version != *product)
+			--start_of_version;
+		start_of_version[1] = '\0';
+		start_of_version += 2;
+		while (start_of_version[0] == ' ' && start_of_version[0] != '\0')
+			++start_of_version;
+		*version = strdup(start_of_version);
+	}
+}
+
+#if defined(_WIN32) || defined(__CYGWIN__)
+probe_info_s *process_ftdi_probe(void)
+{
+	probe_info_s *probe_list = NULL;
+	DWORD ftdiDevCount = 0;
+	char *serial;
+	char *probe_skip = NULL;
+	char *manufacturer;
+	char *product;
+
+	FT_DEVICE_LIST_INFO_NODE *dev_info;
+	if (FT_CreateDeviceInfoList(&ftdiDevCount) == FT_OK) {
+		dev_info = (FT_DEVICE_LIST_INFO_NODE *)malloc(sizeof(FT_DEVICE_LIST_INFO_NODE) * ftdiDevCount);
+		if (dev_info != NULL) {
```

Comment:
Pivoting this to:
```c
if (dev_info == NULL) {
	DEBUG_WARN("process_ftdi_probe: memory allocation failed\n");
	return NULL;
}
```

puts the error reporting much closer to what generates the errors and de-indents the entire function substantially. Please prefer this to the nesting doll approach.